client/packages/react/src/next-ssr/HydrationStreamProvider.tsx

```diff
@@ -0,0 +1,198 @@
+'use client';
+
+// This allows us to send serialized json from the server to the client using 'useServerInsertedHTML'
+// We are allowed to send whatever json we want, however it must be triggered at the same time as the client is recieving new html to display
+// after a suspsense boundary resolves
+
+export const isServer = typeof window === 'undefined' || 'Deno' in globalThis;
+
+import { useServerInsertedHTML } from 'next/navigation.js';
+
+import * as React from 'react';
+import { htmlEscapeJsonString } from './htmlescape.js';
+
+const serializedSymbol = Symbol('serialized');
+
+interface DataTransformer {
+  serialize: (object: any) => any;
+  deserialize: (object: any) => any;
+}
+
+type Serialized<TData> = unknown & {
+  [serializedSymbol]: TData;
+};
+
+interface TypedDataTransformer<TData> {
+  serialize: (obj: TData) => Serialized<TData>;
+  deserialize: (obj: Serialized<TData>) => TData;
+}
+
+interface HydrationStreamContext<TShape> {
+  id: string;
+  stream: {
+    /**
+     * **Server method**
+     * Push a new entry to the stream
+     * Will be ignored on the client
+     */
+    push: (...shape: Array<TShape>) => void;
+  };
+}
+
+export interface HydrationStreamProviderProps<TShape> {
+  children: React.ReactNode;
+  /**
+   * Optional transformer to serialize/deserialize the data
+   * Example devalue, superjson et al
+   */
+  transformer?: DataTransformer;
+  /**
+   * **Client method**
+   * Called in the browser when new entries are received
+   */
+  onEntries: (entries: Array<TShape>) => void;
+  /**
+   * **Server method**
+   * onFlush is called on the server when the cache is flushed
+   */
+  onFlush?: () => Array<TShape>;
+  /**
+   * A nonce that'll allow the inline script to be executed when Content Security Policy is enforced
+   */
+  nonce?: string;
+}
+
+export function createHydrationStreamProvider<TShape>() {
+  const context = React.createContext<HydrationStreamContext<TShape>>(
+    null as any,
+  );
+  /**
+
+   * 1. (Happens on server): `useServerInsertedHTML()` is called **on the server** whenever a `Suspense`-boundary completes
+   *    - This means that we might have some new entries in the cache that needs to be flushed
+   *    - We pass these to the client by inserting a `<script>`-tag where we do `window[id].push(serializedVersionOfCache)`
+   * 2. (Happens in browser) In `useEffect()`:
+   *   - We check if `window[id]` is set to an array and call `push()` on all the entries which will call `onEntries()` with the new entries
```

Comment:
A bit lost with this note. Would get a bit more high level here, and say what the goal is. (I can jump into the code to see the more specific `push` `onEntries` etc.